### PR TITLE
Allocation free batch exporting & CircularBuffer tweaks.

### PR DESCRIPTION
**Pull request by CodeBlanch**
Relates to #1080.

## Changes

* Added `Batch<T>` (a sort of union) which provides for allocation free foreach of `CircularBuffer<T>` instances and provides an allocation free path for passing a single item as a batch and then performing a foreach on that batch.
* Tweaked the `CircularBuffer<T>` algorithm in favor of unblocking writers as quickly as possible.

### CODE REVIEW

File: src/OpenTelemetry/Batch.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+using System.Diagnostics;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry
+{
+    /// <summary>
+    /// Stores a batch of completed <typeparamref name="T"/> objects to be exported.
+    /// </summary>
+    /// <typeparam name="T">The type of object in the <see cref="Batch{T}"/>.</typeparam>
+    public readonly struct Batch<T>
```
Comment (reyang) [inline src/OpenTelemetry/Batch.cs:29]: Is there a way we can mark public APIs as "public for plugin developers only, not for end users"? @cijothomas 
 @eddynaka

Alternative way would be put these in a special namespace such like `OpenTelemetry.ExtensibilityHighVoltageBeCareful`.

File: src/OpenTelemetry/Internal/CircularBuffer.cs [hunk 2]
```diff
@@ -74,24 +66,12 @@ public int Count
...
         /// Gets the number of items added to the <see cref="CircularBuffer{T}"/>.
         /// </summary>
-        public long AddedCount
-        {
-            get
-            {
-                return this.head;
-            }
-        }
+        public long AddedCount => this.head;
```
Comment (reyang) [inline src/OpenTelemetry/Internal/CircularBuffer.cs:69]: I didn't know C# has such syntax sugar 🤣

File: src/OpenTelemetry/Trace/ReentrantExportActivityProcessor.cs [hunk 3]
```diff
@@ -44,8 +44,7 @@ public override void OnEnd(Activity activity)
         {
             try
             {
-                // TODO: avoid heap allocation
-                _ = this.exporter.Export(new[] { activity });
+                _ = this.exporter.Export(new Batch<Activity>(activity));
```
Comment (reyang) [inline src/OpenTelemetry/Trace/ReentrantExportActivityProcessor.cs:47]: 👍👍👍

File: src/OpenTelemetry/Batch.cs (hunk 1 above)
Comment (cijothomas) [reply src/OpenTelemetry/Batch.cs:29]: Some namespace other than `OpenTelemetry` would be good.
like `OpenTelemetry.Extensibility....`?. 
Irrespective of which namespace it is, we still follow the semver guidelines ? (no breaking change unless major version is bumped etc.?

Comment (eddynaka) [reply src/OpenTelemetry/Batch.cs:29]: We could add a <remarks> tag and be clear what are our intentions there.

**Pull request merged**